cpp/include/raft/linalg/cholesky_r1_update.cuh

```diff
@@ -0,0 +1,199 @@
+  // we just update L with the new row. The new Cholesky decomposition will be
+  // calculated as:
+  // L' = [[L_11,    0]
+  //       [L_12, L_22]]
+  // where L_11 is the Cholesky decomposition of A (size [n-1 x n-1]), and
+  // L_12 and L_22 are the new quantities that we need to calculate.
+
+  // We need a workspace in device memory to store a scalar. Additionally, in
+  // CUBLAS_FILL_MODE_LOWER we need space for n-1 floats.
+  const int align = 256;
+  int offset = (uplo == CUBLAS_FILL_MODE_LOWER)
+                 ? raft::alignTo<int>(sizeof(math_t) * (n - 1), align)
+                 : 0;
+  if (workspace == nullptr) {
+    *n_bytes = offset + 1 * sizeof(math_t);
+    return;
+  }
+  math_t *s = reinterpret_cast<math_t *>(((char *)workspace) + offset);
+  math_t *L_22 = L + (n - 1) * ld + n - 1;
+
+  math_t *A_new;
+  math_t *A_row;
+  if (uplo == CUBLAS_FILL_MODE_UPPER) {
+    // A_new is stored as the n-1 th column of L
+    A_new = L + (n - 1) * ld;
+  } else {
+    // If the input is lower triangular, then the new elements of A are stored
+    // as the n-th row of L. Since the matrix is column major, this is non
+    // contiguous. We copy elements from A_row to a contiguous workspace A_new.
+    A_row = L + n - 1;
+    A_new = reinterpret_cast<math_t *>(workspace);
+    CUBLAS_CHECK(raft::linalg::cublasCopy(handle.get_cublas_handle(), n - 1,
+                                          A_row, ld, A_new, 1, stream));
+  }
+  cublasOperation_t op =
+    (uplo == CUBLAS_FILL_MODE_UPPER) ? CUBLAS_OP_T : CUBLAS_OP_N;
+  if (n > 1) {
+    // Calculate L_12 = x by solving equation L_11 x = A_12
+    math_t alpha = 1;
+    CUBLAS_CHECK(raft::linalg::cublastrsm(
+      handle.get_cublas_handle(), CUBLAS_SIDE_LEFT, uplo, op,
+      CUBLAS_DIAG_NON_UNIT, n - 1, 1, &alpha, L, ld, A_new, n - 1, stream));
+
+    // A_new now stores L_12, we calculate s = L_12 * L_12
+    CUBLAS_CHECK(raft::linalg::cublasdot(handle.get_cublas_handle(), n - 1,
+                                         A_new, 1, A_new, 1, s, stream));
+
+    if (uplo == CUBLAS_FILL_MODE_LOWER) {
+      // Copy back the L_12 elements as the n-th row of L
+      CUBLAS_CHECK(raft::linalg::cublasCopy(handle.get_cublas_handle(), n - 1,
+                                            A_new, 1, A_row, ld, stream));
+    }
+  } else {  // n == 1 case
+    CUDA_CHECK(cudaMemsetAsync(s, 0, sizeof(math_t), stream));
+  }
+
+  // L_22 = sqrt(A_22 - L_12 * L_12)
+  raft::linalg::binaryOp(
+    L_22, s, L_22, 1, [] __device__(math_t s, math_t l) { return sqrt(l - s); },
+    stream);
+
+  // Check for numeric error: the above sqrt can have negative argument if the
+  // matrix is not positive definite or very ill conditioned.
+  math_t L_22_host;
+  raft::update_host(&L_22_host, L_22, 1, stream);
+  CUDA_CHECK(cudaStreamSynchronize(stream));
+  if (std::isnan(L_22_host)) {
+    THROW("Error during Cholesky rank one update");
+  }
```

Comment:
nitpick: this could be done via `ASSERT` macro for simplicity, right?

Reply:
Yes, thanks for pointing this out! I have replaced it with assert.